### PR TITLE
Thread-safety for the type metadata cache

### PR DESCRIPTION
**Pull request by dabrahams**
You'll probably want something that avoids the recursive mutex, at least in the common case, but this fixes the immediate correctness issue.

### CODE REVIEW

**Comment by dabrahams** [comment]
@swift-ci Please benchmark

**Comment by dabrahams** [comment]
@swift-ci Please test

**Comment by gottesmm** [comment]
@dabrahams test case?

**Comment by dabrahams** [comment]
@gottesmm I'm trying to reduce it; the one we have is big and involves tensorflow.  tsan blames the metadata cache, though.

**Comment by aschwaighofer** [comment]
Hmm, the cache variable is supposed to be a init once global that is supposed to be thread safe and the Metadata cache itself is supposed to be a concurrent map.

Is the issue on the mac?

Sounds like we have a bug somewhere.

CC @rjmccall

**Comment by compnerd** [comment]
@dabrahams the TSAN results would be useful too

**Comment by dabrahams** [comment]
Here's a reproducer:
```swift
print("*")
import Foundation
protocol P {
  func f(_: Int)
}

struct Inf<X> : P {
  func f(_ x: Int) {
    if x == 0 { return }
    sink(Inf<Self>(), x - 1)
  }
}

@inline(never)
func sink(_ x: Any, _ i: Int) {
  (x as! P).f(i)
}

let N = 20420

#if SERIAL
  for n in 0..<N {
    sink(Inf<()>(), n)
  }
#else
  DispatchQueue.concurrentPerform(iterations: N) { n in
    sink(Inf<()>(), n)
  }
#endif
```
On my MBP, this crashes quickly with N as low as 2042 and no TSAN messages.
On my 96-core linux box it takes longer, but I get:
```
ThreadSanitizer:DEADLYSIGNAL
==114239==ERROR: ThreadSanitizer: SEGV on unknown address 0x6000258ea000 (pc 0x55ebbdbeadb8 bp 0x7f282af781a0 sp 0x7f282af78148 T114283)
==114239==The signal is caused by a WRITE memory access.
ThreadSanitizer:DEADLYSIGNAL
ThreadSanitizer:DEADLYSIGNAL
ThreadSanitizer:DEADLYSIGNAL
ThreadSanitizer:DEADLYSIGNAL
ThreadSanitizer:DEADLYSIGNAL
ThreadSanitizer:DEADLYSIGNAL
ThreadSanitizer:DEADLYSIGNAL
ThreadSanitizer:DEADLYSIGNAL
ThreadSanitizer:DEADLYSIGNAL
ThreadSanitizer: nested bug in the same thread, aborting.
```

Run with `swiftc -sanitize=thread -O main.swift -o /tmp/x && /tmp/x`

**Comment by pschuh** [comment]
in swift_getAssociatedTypeWitness
it is the:
auto *witnessAddr = &((const void* *)wtable)[witnessIndex];
auto witness = *witnessAddr; // <---- here

**Comment by dabrahams** [comment]
Heh, @pschuh reports that my reproducer is tickling a different bug.  He's working on reducing our large test case. @aschwaighofer @gottesmm would you like me to file bugs in Jira for these?

**Comment by dabrahams** [comment]
OK, the real reproducer is attached to https://bugs.swift.org/browse/SR-12760

File: stdlib/public/runtime/Metadata.cpp [hunk 1]
```diff
@@ -684,6 +687,7 @@ MetadataResponse
 swift::swift_getGenericMetadata(MetadataRequest request,
                                 const void * const *arguments,
                                 const TypeContextDescriptor *description) {
+  std::lock_guard<std::recursive_mutex> guard(metadata_mutex);
```
Comment (rjmccall) [inline stdlib/public/runtime/Metadata.cpp:690]: We didn't just forget to add a lock here.  If there's something wrong with our current implementation, we should figure it out, but grabbing a lock at the start of a bunch of runtime functions that are meant to work without locks is not acceptable.

**Review by rjmccall — CHANGES_REQUESTED** [review]

**Comment by rjmccall** [comment]
TSan is right to report about `swift_getAssociatedTypeWitness` / `swift_getAssociatedTypeWitnessSlowImpl`; the loads and stores from the witness table need to be atomic (acquire/release, respectively).

File: stdlib/public/runtime/Metadata.cpp (hunk 1 above)
Comment (dabrahams) [reply stdlib/public/runtime/Metadata.cpp:690]: @rjmccall Hey, we know.  We're just trying to get the problem on everybody's radar.  How these functions are “meant to work” is not immediately obvious from the comments, and I guess some other people around here probably *do* understand that.  If someone would like to explain that, we might be able to attempt an acceptable fix.  We won't be insulted either if this PR is closed and replaced with one that is acceptable. 😉

Comment (rjmccall) [reply stdlib/public/runtime/Metadata.cpp:690]: Well, this is a pretty aggressive way of reporting a bug.

I opened #31768 for the associated-witness issue, which should fix Parker's bug.

Comment (dabrahams) [reply stdlib/public/runtime/Metadata.cpp:690]: Wow, man, no offense intended.  I thought such a bug in the runtime (which I presume to be a security issue) might be important enough that y'all might be interested in a quick fix.  I don't know what you think is aggressive about it.  Just trying to be helpful, here.  Next time I'll stick to Jira.

Comment (rjmccall) [reply stdlib/public/runtime/Metadata.cpp:690]: No offense taken; I'm just trying to explain that this is not, in fact, a working quick fix.

Please CC me on any concurrency bugs you file; we definitely want to treat them as high-priority.

**Comment by dabrahams** [comment]
Closing in favor of #31768

**Pull request closed without merging**